cuda/matrix/csr_kernels.cu

```diff
@@ -246,33 +247,158 @@ void classical_spmv(syn::value_list<int, subwarp_size>,
+                       const matrix::Dense<ValueType>* b,
+                       matrix::Dense<ValueType>* c,
+                       const matrix::Dense<ValueType>* alpha = nullptr,
+                       const matrix::Dense<ValueType>* beta = nullptr)
 {
-    if (a->get_strategy()->get_name() == "load_balance") {
-        components::fill_array(exec, c->get_values(),
-                               c->get_num_stored_elements(), zero<ValueType>());
-        const IndexType nwarps = a->get_num_srow_elements();
-        if (nwarps > 0) {
-            const dim3 csr_block(config::warp_size, warps_in_block, 1);
-            const dim3 csr_grid(ceildiv(nwarps, warps_in_block),
-                                b->get_size()[1]);
+    if (beta) {
+        dense::scale(exec, beta, c);
+    } else {
+        dense::fill(exec, c, zero<ValueType>());
+    }
+    const IndexType nwarps = a->get_num_srow_elements();
+    if (nwarps > 0) {
+        const dim3 csr_block(config::warp_size, warps_in_block, 1);
+        const dim3 csr_grid(ceildiv(nwarps, warps_in_block), b->get_size()[1]);
+        if (alpha) {
             kernel::abstract_spmv<<<csr_grid, csr_block>>>(
                 nwarps, static_cast<IndexType>(a->get_size()[0]),
+                as_cuda_type(alpha->get_const_values()),
                 as_cuda_type(a->get_const_values()), a->get_const_col_idxs(),
                 as_cuda_type(a->get_const_row_ptrs()),
                 as_cuda_type(a->get_const_srow()),
                 as_cuda_type(b->get_const_values()),
                 as_cuda_type(b->get_stride()), as_cuda_type(c->get_values()),
                 as_cuda_type(c->get_stride()));
         } else {
-            GKO_NOT_SUPPORTED(nwarps);
+            kernel::abstract_spmv<<<csr_grid, csr_block>>>(
+                nwarps, static_cast<IndexType>(a->get_size()[0]),
+                as_cuda_type(a->get_const_values()), a->get_const_col_idxs(),
+                as_cuda_type(a->get_const_row_ptrs()),
+                as_cuda_type(a->get_const_srow()),
+                as_cuda_type(b->get_const_values()),
+                as_cuda_type(b->get_stride()), as_cuda_type(c->get_values()),
+                as_cuda_type(c->get_stride()));
         }
+    } else {
+        GKO_NOT_SUPPORTED(nwarps);
+    }
+}
+
+
+template <typename ValueType, typename IndexType>
+bool try_general_sparselib_spmv(std::shared_ptr<const CudaExecutor> exec,
+                                const ValueType* alpha,
+                                const matrix::Csr<ValueType, IndexType>* a,
+                                const matrix::Dense<ValueType>* b,
+                                const ValueType* beta,
+                                matrix::Dense<ValueType>* c)
+{
+    auto handle = exec->get_cusparse_handle();
+#if defined(CUDA_VERSION) && (CUDA_VERSION < 11000)
+    if (b->get_stride() != 1 || c->get_stride() != 1) {
+        return false;
+    }
+
+    auto descr = cusparse::create_mat_descr();
```

Comment:
I think there was some plan on moving this matrix descriptor to an RAII class ?

Reply:
yes, I'll take care of that.

Reply:
Probably in another PR, this one is already huge enough 😄